client/src/components/tables/DataTable.js

```diff
@@ -0,0 +1,206 @@
+ * @returns {JSX.Element}
+ */
+export const DataTable = ({ columns, fetchData, data }) => {
+  const classes = useTableStyles();
+  const [order, setOrder] = useState('asc');
+  const [isLoadingData, setLoadingData] = useState(false);
+  const [isPendingRequests, setIsPendingRequests] = useState(true);
+  const [rows, setRows] = useState([]);
+  const [orderBy, setOrderBy] = useState('name');
+
+  const handleRequestSort = (_, property) => {
+    const isAsc = orderBy === property && order === 'asc';
+    setOrder(isAsc ? 'desc' : 'asc');
+    setOrderBy(property);
+  };
+
+  const loadData = useCallback(async () => {
+    setLoadingData(true);
+    const rowsData = await fetchData(columns);
+
+    setRows(rowsData);
+    setIsPendingRequests(rowsData.length > 0);
+    setLoadingData(false);
+  }, [setRows, fetchData, columns]);
+
+  const sort = (array) => sortObjects(array, orderBy, order);
+
+  useEffect(() => {
+    if (data) {
+      setRows(data);
+    } else {
+      loadData();
+    }
+  }, [data, loadData]);
+
+  const types = Object.fromEntries(columns.map((column) => [column.id, column.type]));
+  const modalList = columns
+    .filter((column) => column.button && column.button.modal)
+    .map((column) => ({ id: column.id, modal: column.button.modal }));
+
+  const [modalsState, setModalsState] = useState(
+    Object.fromEntries(
+      columns
+        .filter((column) => column.button && column.button.modal)
+        .map((column) => [column.id, { open: false, content: {} }])
+    )
+  );
+
+  /**
+   * @param {string} id      ID of the row to to open the modal for.
+   * @param {object} content Content to pass to the opened modal.
+   */
+  const openModal = (id, content) => {
+    const newState = { ...modalsState };
+    newState[id].open = true;
+    newState[id].content = content;
+    setModalsState(newState);
+  };
+
+  /**
+   * @param {string} id      ID of the row to to close the modal for.
+   */
+  const closeModal = (id) => {
+    const newState = { ...modalsState };
+    newState[id].open = false;
+    setModalsState(newState);
+  };
+
+  const modalSubmit = (id) => {
+    closeModal(id);
+    loadData();
+  };
+
+  const buttonConfigs = Object.fromEntries(
+    columns.map((column) => {
+      if (column.button && !column.button.callback && column.button.modal) {
+        column.button.callback = (row) => openModal(column.id, row);
+      }
+      return [column.id, column.button];
+    })
+  );
+
+  return (
+    <>
+      {modalList.map((modal) => (
+        <modal.modal
+          key={modal.id}
+          open={modalsState[modal.id].open}
+          onSubmit={() => modalSubmit(modal.id)}
+          onClose={() => closeModal(modal.id)}
+          content={modalsState[modal.id].content}
+        />
+      ))}
+
+      <Grid
+        container
+        alignContent='flex-start'
+        justify='flex-start'
+        alignItems='center'
+        direction='row'
+      >
+        {isPendingRequests && (
```

Comment:
`isPendingRequests` feels like a confusing name, but I'm not sure what a better one is. Does it refer to a loading or finished state? It seems that it refers to finished, but the var name seems counter to that.

Reply:
Yeah, that's a variable name borrowed from some of the older table components. I'll make it more readable.